src/libprojectM/MilkdropPresetFactory/Expr.cpp

```diff
@@ -24,156 +24,394 @@
+    arg_types.push_back(llvm::PointerType::get(jitx.floatType,1)); // float *
+    auto prefun_type = llvm::FunctionType::get(jitx.floatType, arg_types, false);
+    auto prefun_ptr_type = llvm::PointerType::get(prefun_type,1);
+    llvm::Constant *fn_const = llvm::ConstantInt::get(llvm::Type::getInt64Ty(jitx.context), (uint64_t)func_ptr);
+    auto function_ptr = llvm::ConstantExpr::getIntToPtr(fn_const , prefun_ptr_type);
+
+    std::vector<llvm::Value *> args;
+    args.push_back(array);
+    return jitx.builder.CreateCall(function_ptr, args);
+}
+#endif
+
+
 class PrefunExprOne : public PrefunExpr
 {
-	float eval ( int mesh_i, int mesh_j )
+public:
+    PrefunExprOne(Func *function, Expr **expr_list_) : PrefunExpr(function,expr_list_) {}
+    float eval ( int mesh_i, int mesh_j )
+    {
+        float val = expr_list[0]->eval ( mesh_i, mesh_j );
+        return (func_ptr)(&val);
+    }
+};
+
+
+class IfAboveExpr : public PrefunExpr
+{
+public:
+	IfAboveExpr(Expr *a, Expr *b, Expr *t, Expr *e) : PrefunExpr()
 	{
-		float val = expr_list[0]->eval ( mesh_i, mesh_j );
-		return (func_ptr)(&val);
+        num_args = 4;
+		expr_list = (Expr **)malloc(num_args*sizeof(Expr *));
+		expr_list[0] = a;
+		expr_list[1] = b;
+		expr_list[2] = t;
+		expr_list[3] = e;
+	}
+	float eval ( int mesh_i, int mesh_j ) override
+	{
+		// see above_wrapper(), below_wrapper()
+		float aval = expr_list[0]->eval(mesh_i,mesh_j);
+		float bval = expr_list[1]->eval(mesh_i,mesh_j);
+		if (aval > bval)
+			return expr_list[2]->eval(mesh_i,mesh_j);
+		else
+			return expr_list[3]->eval(mesh_i,mesh_j);
+	}
+#if HAVE_LLVM
+    llvm::Value *_llvm(JitContext &jitx) override
+    {
+        llvm::Value *aValue = Expr::llvm(jitx, expr_list[0]);
+        llvm::Value *bValue = Expr::llvm(jitx, expr_list[1]);
+        if (nullptr == aValue || nullptr == bValue)
+            return nullptr;
+        llvm::Value *condition = jitx.builder.CreateFCmpUGT(aValue, bValue);
+        jitx.StartTernary(condition);
+        jitx.withThen();
+        llvm::Value *thenValue = Expr::llvm(jitx, expr_list[2]);
+        jitx.withElse();
+        llvm::Value *elseValue = Expr::llvm(jitx, expr_list[3]);
+        if (nullptr == thenValue || nullptr == elseValue)
+            return nullptr;
+        return jitx.FinishTernary(thenValue, elseValue);
+    }
+#endif
+};
+
+
+class IfEqualExpr : public PrefunExpr
+{
+public:
+	IfEqualExpr(Expr *a, Expr *b, Expr *t, Expr *e) : PrefunExpr()
+	{
+        num_args = 4;
+        expr_list = (Expr **)malloc(num_args*sizeof(Expr *));
+		expr_list[0] = a;
+		expr_list[1] = b;
+		expr_list[2] = t;
+		expr_list[3] = e;
 	}
+	float eval ( int mesh_i, int mesh_j ) override
+	{
+		// see above_wrapper(), below_wrapper()
+		float aval = expr_list[0]->eval(mesh_i,mesh_j);
+		float bval = expr_list[1]->eval(mesh_i,mesh_j);
+		if (aval == bval)
+			return expr_list[2]->eval(mesh_i,mesh_j);
+		else
+			return expr_list[3]->eval(mesh_i,mesh_j);
+	}
+#if HAVE_LLVM
```

Comment:
is there a way to avoid adding a ton of conditional compilation checks, like maybe making a ::JIT subclass for each expression and using those if HAVE_LLVM?